quipucords/scanner/tasks.py

```diff
@@ -69,32 +70,79 @@ def get_task_runner_class(source_type, scan_type) -> type[ScanTaskRunner]:
     return getattr(get_scanner(source_type), runner_class_name)
 
 
-@celery.shared_task(bind=True)
-def celery_run_task_runner(
-    self: celery.Task, scan_task_id: int, source_type: str, scan_type: str
-) -> tuple[bool, int, str]:
+def set_scan_task_failure_on_exception(func):
     """
-    Wrap _celery_run_task_runner to call it as an async Celery task.
+    Set the ScanTask's FAILURE status when handling unexpected exceptions.
+
+    This decorator assumes `scan_task_id` will be the second argument or a keyword
+    argument of the function being wrapped.
 
-    Also wrap the call with try-except to handle any unexpected exceptions
-    and update the task's status to "failed" if any exception was raised.
+    Todo: Maybe merge/consolidate this with set_scan_job_failure_on_exception?
+    If we merge these two decorators, we probably need a clever way to determine
+    whether the relevant object to "fail" is a ScanJob or a ScanTask.
     """
-    try:
-        return _celery_run_task_runner(self, scan_task_id, source_type, scan_type)
-    except Exception:  # noqa: BLE001
-        logger.exception(
-            f"Unexpected exception in celery_run_task_runner (ScanTask {scan_task_id})"
-        )
 
-    try:
-        scan_task = ScanTask.objects.get(id=scan_task_id)
-        scan_task.status_fail(
-            f"Unexpected exception in celery_run_task_runner (ScanTask {scan_task_id})"
-        )
-    except Exception:  # noqa: BLE001
-        logger.exception(
-            f"Failed to record celery_run_task_runner failure (ScanTask {scan_task_id})"
-        )
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        scan_task_id = args[1] if len(args) > 1 else kwargs.get("scan_task_id", None)
```

Comment:
I don't love the idea of inferring the argument by position as default. I'd consume it as a kwarg and force decorated functions to use it as a kwarg
you probably know how to do this, but just in case you don't, here's how to:

```python
def some_func(*, keyword_only_arg):
    ...
    ````

Reply:
and the reason I don't like the positional argument is because someone might change it in the future and it could become a bit hard to debug why things weren't working as expected.

Reply:
I like the suggestion, but I'm not sure if the `*` keyword enforcer works safely with the way Celery task are defined and invoked, though, or at least when we set `bind=True` in `@celery.shared_task`. Using `bind` defers some calling logic to Celery internals to set an _additional_ argument at the start of the function calls, and that's why we have `self` on some task functions like this:

```python
@celery.shared_task(bind=True)
@set_scan_task_failure_on_exception
def fingerprint(self: celery.Task, scan_task_id: int) -> tuple[bool, int, str]:
    ...
```
We use that `self` to interrogate things about the task itself, such as its ID so we can check if it's been canceled.

Does that `self` work with `*`? I cannot tell yet. I am trying to parse through the Celery and Kombu code to try to understand how its `bind` argument actually works, but it quickly gets complicated.  😕 

I have at least locally (and maybe only temporarily) replaced that ugly one-liner with this as a potential safety check if `*` doesn't work as expected:

```python
if not (scan_task_id := kwargs.get("scan_task_id", None)):
    logger.warning(
        f"Missing scan_task_id kwarg for decorated function {func.__name__}"
    )
```

Reply:
Further testing with `*` reveals things to be… still complicated. For example, having this task defined:

```python
@celery.shared_task(bind=True)
@set_scan_job_failure_on_exception
def finalize_scan(self: celery.Task, *, scan_job_id: int):
    """Wrap _finalize_scan to call it as an async Celery task."""
    return _finalize_scan(self=self, scan_job_id=scan_job_id)
```

One would expect `scan_job_id` to be a required keyword argument, right? However, we never invoke this function directly. Because it's a Celery task, we indirectly invoke it by requesting it to be run through `finalize_scan.delay(...)`, and that `delay` function does not enforce the `*` magic, as far as I can tell. So, the caller can "successfully" call `finalize_scan` without `scan_job_id` as a keyword argument, and it only fails later (I think) when handled by the Celery worker. This means we need to be _very careful_ in how we write unit tests where we don't fully execute the Celery tasks (or use `CELERY_TASK_ALWAYS_EAGER`). Totally workable, but something to keep in mind.